src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs

```diff
@@ -47,7 +47,7 @@ public unsafe void Append(string s)
             Append(s, 0, s.Length);
         }
 
-        public void Append(StringSegment segment)
+        public void Append(in StringSegment segment)
```

Comment:
This is breaking, it changes the signature to be the equivalent of `ref`, but with an additional readonly attribute for compiler disambiguation (the ref changes the signature, the attribute does not)

Reply:
If an additional method was added, which would the compiler prefer, as to the caller they are the same sig?
```
public void Append(StringSegment segment)
public void Append(in StringSegment segment)
```

Reply:
Seems it gets confused
![image](https://user-images.githubusercontent.com/1142958/35784126-6381fac6-0a0b-11e8-9149-e6872eba5143.png)

Reply:
The "confusion"  is getting resolved in 7.3, iirc.

The byval will always win by default and you will need to explicitly specify `in` to call the other overload.

Reply:
If you can call the `in` without `in` when there is only one overload; then it should clearly also win when its a `readonly struct` as there is no disadvantage to it winning only an advantage?

Reply:
That depends on the size of the struct, the underlying ABI, and whether or not you want different overload resolution for a normal struct vs a readonly struct

Reply:
Differing read only vs normal can see as perhaps weird.

As to size and ABI; overload is going to be the same regardless; so really its a question of whether the overload should have been written in the first place; since it is there must be a reason (unless you are ifdefing it)

Reply:
And if you are ifdefing it, you just ifdef out the overload that's not suitable for the platform

Reply:
I was primarily indicating there may be some scenarios where it isn't always advantageous.

> As to size and ABI; overload is going to be the same regardless; so really that's a question of whether the overload should have been written in the first place; since it has there must be a reason (unless you are ifdefing it)

Yes, and defaulting to the `byval` overload is, IMO, the "better" default (from a language perspective). It means that we don't need another keyword to indicate the `byval` overload should be called, it means existing code will continue calling the same overload they did before, and that they can explicitly opt-in to calling the new overload if needed/where desired.